NFTs filters selector

kkosiorowska opened the pull request:
Closes #2601 

This PR allows filtering NFTs. A separate panel was created where the user can set the filtering parameters. 

Possible filtering options for collections:
- show/hide chosen accounts
- show/hide chosen collections
- sort type by price, date and numbers

## UI
https://user-images.githubusercontent.com/23117945/207604752-8ace7fb4-826c-4214-a487-8a4b6969cd1b.mov



## To Test
- [ ] check if the user gets a message when there are no matching NFTs
- [ ] check that sorting works for all types
- [ ] check that the account filter is working
- [ ] check that the collections filter is working
- [ ] check if the header changes (NFTs count, Badges count, total value) - NFTs, Portfolio page
- [ ] check that the filter icon isn't visible when the user doesn't have NFTS

## Testing Env

```
SUPPORT_NFT_TAB=true
```

Latest build: [extension-builds-2764](https://github.com/tallyhowallet/extension/suites/10031794118/artifacts/487762171) (as of Fri, 23 Dec 2022 15:30:03 GMT).

greg-nagy commented:
nope

VladUXUI commented:
This is looking veeeeeery nice @kkosiorowska !! 

Noticed a couple of things with filters tho:

- [ ]  If estimated total value is a big number, it pushes the filter out of view
<img width="400" alt="image" src="https://user-images.githubusercontent.com/80846886/207835176-759ca872-47e6-457c-ae64-f7a05d55bba6.png">

- [ ] Design for radio buttons seems to be wrong, this is our design:
<img width="296" alt="image" src="https://user-images.githubusercontent.com/80846886/207835635-71461104-eb8d-4c49-88f4-cbb43d8cc3cd.png">
Right now it looks like this:
<img width="162" alt="image" src="https://user-images.githubusercontent.com/80846886/207835742-8d09811b-3574-4445-972a-cbf2a19d6fbd.png">

VladUXUI commented:
Another thing i noticed, and i didn't realise this until i started using it is that filters should be a bit re-arranged and collections filter list should be filtered by accounts active. 
Like this:
<img width="709" alt="image" src="https://user-images.githubusercontent.com/80846886/207836663-849fedf1-dba3-4777-a754-dbc367fc91c5.png">

This is not a blocker, and if it takes time to do, i can create a follow-up ticket

jagodarybacka commented on background/redux-slices/nfts_update.ts:
```diff
@@ -28,7 +28,20 @@ export type NFTWithCollection = {
+  id: string
+  name: string
+  isEnabled: boolean
+  avatarURL?: string
```
can we use a `thumbnailURL` name? We already have `thumbnail` and `preview` so I don't want to add `avatar` 😂

jagodarybacka commented on ui/components/NFTS_update/NFTListPortfolio.tsx:
```diff
@@ -1,10 +1,61 @@
+              ? t("units.collection_other")
+              : t("units.badge_other")}
+          </h2>
+          <NFTList collections={collections} />
```
`expandBadgesCollections` should be added if this is for badges tab

jagodarybacka commented on ui/components/Shared/SharedRadio.tsx:
```diff
@@ -0,0 +1,62 @@
+  return (
+    <div className="radio">
+      <input
+        id="radio"
```
You should have 1 unique id in the DOM at any given time - that's why clicking the label of the radio button is not working and hovering them is funky

jagodarybacka commented on ui/components/Shared/SharedRadio.tsx:
```diff
@@ -0,0 +1,62 @@
+      <input
+        id="radio"
+        type="radio"
+        name={name || "radio"}
```
hmm I'm not sure if we need this name - what is it used for?

kkosiorowska replied:
`name` is used to ensure that only one radio button can be selected from the group.

jagodarybacka commented on ui/components/Shared/SharedSlideUpMenuPanel.tsx:
```diff
@@ -0,0 +1,37 @@
+  children?: ReactNode
+}
+
+export default function SharedSlideUpMenuPanel({
```
I have seen it in the other [PR](https://github.com/tallycash/extension/pull/1840/files#r1045769425) - can we make it reusable so both places can use your implementation?

kkosiorowska replied:
I added a `style` property for the header. It will be needed to change the font size of the header. The size of SharedSlideUpMenu has to be set up using `custom` and `customSize` property. 

![Screenshot 2022-12-19 at 17 56 54](https://user-images.githubusercontent.com/23117945/208480325-dfc3febc-6a4b-41c3-8ba7-7a00e3d38f84.png)

jagodarybacka commented on ui/components/NFTS_update/NFTsHeader.tsx:
```diff
@@ -125,8 +156,7 @@ export default function NFTsHeader(): ReactElement {
           align-items: center;
-          justify-content: space-evenly;
-          width: 100%;
+          gap: 16px;
```
Can we not set this with a specific value? I wanted to allow as much text as possible to be included in one line so "Collection", "NFTs" and "Badges" are not moved to the second line when you check out accounts that have a lot of NFTs

jagodarybacka reviewed (commented):
I haven't tested the app as much yet, will do but quick Q:
- should stats be affected by selected filters? Total value, number of NFTs and badges?

jagodarybacka reviewed (commented):
In-app testing:

- removing account is not updating filters

- switching on-off collections is switching their places (may be linked to the POAPs as they are handled differently 🤔 )

https://user-images.githubusercontent.com/20949277/207861903-62101451-8abf-4136-afb0-5c60a8b18938.mov

- seems like we know the price for this NFT but we don't show it?

https://user-images.githubusercontent.com/20949277/207863013-bf4c4e9d-0da1-4147-af39-2d786e98abff.mov

- should stats on the Portfolio page be affected by filters?

https://user-images.githubusercontent.com/20949277/207863532-493c5219-5680-477e-885d-40190f0c8678.mov

- if the account is turned off with a switch - can we hide or disable in filters all collections that are owned by this disabled account but are not owned by other accounts?

kkosiorowska commented:
Done :+1:
> * [ ]   If estimated total value is a big number, it pushes the filter out of view

![Screenshot 2022-12-20 at 15 03 51](https://user-images.githubusercontent.com/23117945/208685232-f9b9fd9b-3e2e-4034-be34-e158a13d3aec.png)

> * [ ]  Design for radio buttons seems to be wrong, this is our design:

![Screenshot 2022-12-20 at 15 04 00](https://user-images.githubusercontent.com/23117945/208685349-943aa135-7b45-4f88-a14c-43ed41b24fef.png)

kkosiorowska commented:
@jagodarybacka 
After improvements:

> * removing account is not updating filters

<img width="80" alt="23117945/208696883-b84cd959-b6b3-44ab-a72f-7e1ec94aada4" src="https://user-images.githubusercontent.com/23117945/208696883-b84cd959-b6b3-44ab-a72f-7e1ec94aada4.mov">

> * switching on-off collections is switching their places (may be linked to the POAPs as they are handled differently 🤔 )

It should be fixed after the latest updates.
https://user-images.githubusercontent.com/23117945/208754986-829a9fdb-8cdc-4e38-b61d-c2ceec84c6a0.mov

> * seems like we know the price for this NFT but we don't show it?

Sorting works. The issue is that we didn't show the price of the NFTs when the value is 0.

https://user-images.githubusercontent.com/23117945/208851034-d01df735-9dff-4338-909b-199357f7a191.mov


> * should stats on the Portfolio page be affected by filters?

I fixed it. Portfolio page will display statistics for all nfts.
https://user-images.githubusercontent.com/23117945/208698194-e6346dba-1e25-45a4-bace-cd876c63d048.mov

> * if the account is turned off with a switch - can we hide or disable in filters all collections that are owned by this disabled account but are not owned by other accounts?

Let's do it.  For the moment, I added a spinner when we change the account filter. It doesn't look great.  @VladUXUI  Do you have any suggestions?

https://user-images.githubusercontent.com/23117945/208714082-a1674c4f-631d-4218-a348-3a5b65e720e6.mov

VladUXUI commented:
>Let's do it. For the moment, I added a spinner when we change the account filter. It doesn't look great. @VladUXUI Do you have any suggestions?

I don't think the issue is the spinner, i think that is fine (alternatively we can use the corgy loader gif).
I think the issue is that it resets the page to top, and that makes it jump and it's disturbing.
Any chance we can keep position of page the same?

kkosiorowska commented:
> I don't think the issue is the spinner, i think that is fine (alternatively we can use the corgy loader gif). I think the issue is that it resets the page to top, and that makes it jump and it's disturbing. Any chance we can keep position of page the same?

I will try to improve it

kkosiorowska commented on ui/components/NFTS_update/Filters/FilterList.tsx:
```diff
@@ -0,0 +1,57 @@
+      setItems(filters)
+      const timeout = setTimeout(() => {
+        setIsEmpty(!filters.length)
+      }, 250)
```
Let's make it simple.

kkosiorowska commented:
Animations after changes:

https://user-images.githubusercontent.com/23117945/209347112-1c366276-9ad4-4292-b620-26edbf304828.mov

jagodarybacka reviewed (approved):
Works ok, we can iterate based on this.
Possible improvements:
- list of collections filters - loading animation (cc @VladUXUI would be nice to do design review)
- POAPs are sometimes visible in collections filters even if the account has no POAPs
